Update Vulkan, SPIRV, etc to v1.4.321.0

tophyr opened the pull request:
<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/. -->

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

<!-- If this PR adds a new port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [ ] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [ ] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [ ] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ ] The license declaration in `vcpkg.json` matches what upstream says.
- [ ] The installed as the "copyright" file matches what upstream says.
- [ ] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is in the new port's versions file.
- [ ] Only one version is added to each modified port's versions file.

END OF NEW PORT CHECKLIST (delete this line) -->

tophyr committed:
Update Vulkan, SPIRV, etc to v1.4.321.0

tophyr commented:
cc @sandercox

tophyr commented:
Fixes incoming.

tophyr commented:
and finally cc @BillyONeal since I see you approving PRs so I assume you have some familiarity with the CI. Thanks in advance.

tophyr commented:
@dg0yt looks like i'm blocked behind #46569

dg0yt commented:
Checked the logs. The android build is never attempted because the linux build fails:
~~~
ggml_vulkan: Generating and compiling shaders to SPIR-V
cannot compile get_rows_iq1_m

/mnt/vcpkg-ci/installed/x64-linux/tools/shaderc/glslc -fshader-stage=compute --target-env=vulkan1.2 -O /mnt/vcpkg-ci/b/ggml/src/d5969d6f49-e3532476b4.clean/src/ggml-vulkan/vulkan-shaders/get_rows_quant.comp -o /mnt/vcpkg-ci/b/ggml/x64-linux-dbg/src/ggml-vulkan/vulkan-shaders.spv/get_rows_iq1_m.spv -DB_TYPE=int -DDATA_A_IQ1_M=1 -DD_TYPE=float16_t -DFLOAT_TYPE=float 

shaderc: internal error: compilation succeeded but failed to optimize: [VUID-StandaloneSpirv-None-10684] Invalid explicit layout decorations on type for operand '1230[%1230]'
  %scales = OpVariable %_ptr_Function__arr_ushort_uint_4 Function



cannot compile get_rows_iq1_m_f32

/mnt/vcpkg-ci/installed/x64-linux/tools/shaderc/glslc -fshader-stage=compute --target-env=vulkan1.2 -O /mnt/vcpkg-ci/b/ggml/src/d5969d6f49-e3532476b4.clean/src/ggml-vulkan/vulkan-shaders/get_rows_quant.comp -o /mnt/vcpkg-ci/b/ggml/x64-linux-dbg/src/ggml-vulkan/vulkan-shaders.spv/get_rows_iq1_m_f32.spv -DB_TYPE=int -DDATA_A_IQ1_M=1 -DD_TYPE=float -DFLOAT_TYPE=float 

shaderc: internal error: compilation succeeded but failed to optimize: [VUID-StandaloneSpirv-None-10684] Invalid explicit layout decorations on type for operand '1230[%1230]'
  %scales = OpVariable %_ptr_Function__arr_ushort_uint_4 Function

~~~
shaderc isn't updated in this PR, but it depends on spirv-tools.

tophyr committed:
update shaderc as well

tophyr committed:
patch ggml until https://github.com/ggml-org/ggml/pull/1317

dg0yt commented:
> @dg0yt oh, interesting. I had been assuming it was `ggml` that needed updating (which I can see you're working on) but what you're saying is that `shaderc` is the root cause?

No, I'm not saying anything about the cause. I just try to elaborate what has changes since the last succesful build of `ggml[vulkan]`, #45756.

tophyr committed:
fix spirv-tools

tophyr committed:
move to qt6 for vulkantools 1.4.321 and disallow all static configs

the vcpkg port already prohibits staticcrt, but the upstream vulkantools project explicitly sets build flags to link against the static runtime when it detects that qt is linked statically. so, vcpkg already prohibits staticcrt and upstream effectively prohibits static link + dynamic crt. so, we just disable static compilation altogether.

tophyr committed:
fix versions

tophyr commented:
> This is https://github.com/microsoft/vcpkg/pull/46058, and it creates installation order problems

I wondered about that - this PR has grown to touch a fair bit more ports than I had expected to or would ideally like to. Advice? I'd _like_ to not be blocked behind anything. IIRC it actually built fine without the move to Qt6, but since the VulkanTools docs explicitly name 6 (not 5) as the dependency for 1.4.321 I figured better to do that.

> > and disallow all static configs
>
> This doesn't seem right.

the commit message itself lays out the logic: the existing vcpkg port disallows `staticcrt`, ie `-static`. (i don't know why the port had that configuration - @moritz-h any insight?) however, the VulkanTools cmake file explicitly _forces_ `/MT` (static crt) when it detects that the Qt it is linking is a static library (ie, when qt is build statically - as it is in `-static-md` - then the VulkanTools build ignores the `-static-md` configuration and explicitly builds itself as if with `-static`.)

tophyr commented:
We'll need to patch out https://github.com/LunarG/VulkanTools/commit/0a2dad5cb501e1d201c62e6f9f6ba86079629829 if we go the "force the triplet that vcpkg wants" route. It's late for me; I'll see if I can think through the implications of that tomorrow.

dg0yt commented:
So `!staticcrt` was added together with `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)`. The latter is still in the port file. Today we shall use `!(static & staticcrt)` for such packages.

dg0yt commented:
IMO the expression should be updated AND the spot should be patched. I even asked upstream about the change - upstream's PR refers to their "local version of Qt" which probably doesn't match our build. (https://github.com/LunarG/VulkanTools/pull/2453#issuecomment-3134946293)

dg0yt commented:
Assuming that `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` shall not be removed, I really mean there shall be `!(static & staticcrt)`:
https://learn.microsoft.com/en-us/vcpkg/maintainers/functions/vcpkg_check_linkage#only_dynamic_library

In vcpkg CI, this will only exclude the `x64-windows-static` triplet (as before with plain `staticcrt`).
`x64-windows-static-md` has a static Qt but with dynamic CRT. And so do all the osx, linux, android triplets in CI. A plain `!static` would disable the port for all of them.

dg0yt commented:
#46058 was merged, so lunarg-vulkantools is now using Qt6. qtwebengine fixed.
